sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,493 @@
+
+1. `(fetch-entry map-name key-tuple)` - This fetches the value
+  associated with a given key in the map, or returns `'null` if there
+  is no such value.
+2. `(set-entry! map-name key-tuple value-tuple)` - This will set the
+  value of `key-tuple` in the data map
+3. `(insert-entry! map-name key-tuple value-tuple)` - This will set
+  the value of `key-tuple` in the data map if and only if an entry
+  does not already exist.
+4. `(delete-entry! map-name key-tuple)` - This will delete `key-tuple`
+   from the data map
+
+We chose to use data maps as opposed to other data structures for two
+reasons:
+
+1. The simplicity of data maps allows for both a simple
+implementation within the VM, and easier reasoning about
+transactions. By inspecting a given transaction definition, it is
+clear which maps will be modified and even within those maps, which
+keys are affected by a given transaction.
+2. The interface of data maps ensures that the return types of map
+operations are _fixed length_, which is a requirement for static
+analysis of smart contracts' runtime, costs, and other properties.
+
+A smart contract defines the data schema of a data map with the
+`defmap` call. The `defmap` function may only be called in the
+top-level of the smart-contract (similar to `define`). This
+function accepts a name for the map, and a definition of the structure
+of the key and value types. Each of these is a list of `(name, type)`
+pairs, and they specify the input and output type of `fetch-entry`.
+Types are either the values `'principal`, `'integer`, `'bool` or
+the output of a call to `(buffer n)`, which defines an n-byte
+fixed-length buffer. 
+
+### Record Type Syntax
+
+To support the use of _named_ fields in keys and values, our language
+allows the construction of named tuples using a function `(tuple ...)`,
+e.g.,
+
+```
+(define imaginary-number-a (tuple #real 1 #i 2))
+(define imaginary-number-b (tuple #real 2 #i 3))
+
+```
+
+This allows for creating named tuples on the fly, which is useful for
+data maps where the keys and values are themselves named tuples. To
+access a named value of a given tuple, the function `(get #name
+tuple)` will return that item from the tuple.
+
+### Reading from Other Smart Contracts
+
+While a smart contract may not _modify_ other smart contracts' data
+directly, it _can_ read data stored in those smart contracts' maps.
+(Note: this does not alter any confidentiality guarantees of the smart
+contracting language. All data in the smart contracts is inherently
+public, and will be readable through querying the underlying database
+in any case.) In order to do so, a contract may use the
+`(fetch-contract-entry)` function, which behaves identically to
+`(fetch-entry)`, though it accepts a contract principal as an argument
+in addition to the map name. Just as with the `(contract-call)` function,
+the map name and contract principal arguments must be constants, specified
+at the time of publishing.
```

Comment:
For completeness, can you add the signature of `fetch-contract-entry` and an example?  Thanks!

Reply:
Added